// floggy/dist-esm/settings.js
import { chalk } from './chalk';
import * as Filter from './filter';
import * as Level from './level';
import * as Output from './output';
import { casesHandled, omitUndefinedKeys, parseFromEnvironment } from './utils';
/**
 * Process data setting input.
 */
export function processSettingInputData(data, previous) {
    if (!previous) {
        return {
            ...defaultSettingData(),
            ...omitUndefinedKeys(data)
        };
    }
    return {
        ...previous,
        ...omitUndefinedKeys(data !== null && data !== void 0 ? data : {})
    };
}
function defaultSettingData() {
    var _a;
    if (((_a = process.env) === null || _a === void 0 ? void 0 : _a.NODE_ENV) === 'production') {
        return {
            hostname: true,
            pid: true,
            time: true
        };
    }
    else {
        return {
            hostname: false,
            pid: false,
            time: false
        };
    }
}
/**
 * Process pretty setting input.
 */
export function processSettingInputPretty(pretty, previous) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    // todo no semantic to "unset back to default"
    // consider using `null` for that purpose...
    const color = (_b = (_a = (typeof pretty === 'object' ? pretty.color : undefined)) !== null && _a !== void 0 ? _a : previous === null || previous === void 0 ? void 0 : previous.color) !== null && _b !== void 0 ? _b : true;
    const enabled = (_d = (_c = (typeof pretty === 'object' ? pretty.enabled : undefined)) !== null && _c !== void 0 ? _c : previous === null || previous === void 0 ? void 0 : previous.enabled) !== null && _d !== void 0 ? _d : 
    // todo nice is-defined-but-parse-error feedback
    (((_f = (_e = process.env) === null || _e === void 0 ? void 0 : _e.LOG_PRETTY) === null || _f === void 0 ? void 0 : _f.toLowerCase()) === 'true'
        ? true
        : ((_h = (_g = process.env) === null || _g === void 0 ? void 0 : _g.LOG_PRETTY) === null || _h === void 0 ? void 0 : _h.toLowerCase()) === 'false'
            ? false
            : (_j = process.stdout) === null || _j === void 0 ? void 0 : _j.isTTY);
    const levelLabel = (_l = (_k = (typeof pretty === 'object' ? pretty.levelLabel : undefined)) !== null && _k !== void 0 ? _k : previous === null || previous === void 0 ? void 0 : previous.levelLabel) !== null && _l !== void 0 ? _l : false;
    const timeDiff = (_o = (_m = (typeof pretty === 'object' ? pretty.timeDiff : undefined)) !== null && _m !== void 0 ? _m : previous === null || previous === void 0 ? void 0 : previous.timeDiff) !== null && _o !== void 0 ? _o : true;
    if (pretty === undefined) {
        return { enabled, color, levelLabel, timeDiff };
    }
    if (pretty === true) {
        return { enabled: true, color, levelLabel, timeDiff };
    }
    if (pretty === false) {
        return { enabled: false, color, levelLabel, timeDiff };
    }
    if (typeof pretty === 'object') {
        return { enabled, color, levelLabel, timeDiff };
    }
    casesHandled(pretty);
}
export function create(opts) {
    var _a, _b;
    const state = {
        pretty: processSettingInputPretty(opts === null || opts === void 0 ? void 0 : opts.pretty, null),
        filter: !opts || !opts.filter || opts.filter === '' || Object.keys(opts).length === 0
            ? defaultFilterSetting()
            : processSettingInputFilter(opts.filter, null),
        output: (_a = opts === null || opts === void 0 ? void 0 : opts.output) !== null && _a !== void 0 ? _a : Output.defaultOutput,
        data: processSettingInputData((_b = opts === null || opts === void 0 ? void 0 : opts.data) !== null && _b !== void 0 ? _b : {}, null)
    };
    const settings = ((newSettings) => {
        if (newSettings.output) {
            // @ts-expect-error ...
            settings.output = newSettings.output;
        }
        if ('pretty' in newSettings) {
            // @ts-expect-error ...
            settings.pretty = processSettingInputPretty(newSettings.pretty, settings.pretty);
            // Sync chalk
            // Assume true color support, not doing all that -> https://github.com/chalk/chalk#256-and-truecolor-color-support
            chalk.level = settings.pretty.color ? 3 : 0;
        }
        if ('data' in newSettings) {
            // @ts-expect-error ...
            settings.data = processSettingInputData(newSettings.data, settings.data);
        }
        if (newSettings.filter && Object.keys(newSettings).length > 0) {
            // @ts-expect-error ...
            settings.filter = processSettingInputFilter(newSettings.filter, settings.filter);
        }
    });
    Object.assign(settings, state);
    return settings;
}
export function processSettingInputFilter(newSettingsFilter, prev) {
    var _a;
    if (!prev)
        prev = defaultFilterSetting();
    newSettingsFilter =
        typeof newSettingsFilter === 'string' ? { pattern: newSettingsFilter } : newSettingsFilter;
    const pattern = (_a = newSettingsFilter.pattern) !== null && _a !== void 0 ? _a : prev.originalInput;
    const defaults = (newSettingsFilter === null || newSettingsFilter === void 0 ? void 0 : newSettingsFilter.level)
        ? { level: { value: newSettingsFilter.level, comp: 'gte' } }
        : prev.defaults;
    const patterns = Filter.processLogFilterInput(defaults, pattern) || Filter.parseUnsafe(defaults, '*');
    return {
        defaults,
        originalInput: pattern,
        patterns
    };
}
export function defaultFilterSetting() {
    var _a, _b, _c, _d;
    let level;
    if ((_a = process.env) === null || _a === void 0 ? void 0 : _a.LOG_LEVEL) {
        level = parseFromEnvironment('LOG_LEVEL', Level.parser);
    }
    else {
        level = ((_b = process.env) === null || _b === void 0 ? void 0 : _b.NODE_ENV) === 'production' ? Level.LEVELS.info.label : Level.LEVELS.debug.label;
    }
    let originalInput;
    let patterns;
    if ((_c = process.env) === null || _c === void 0 ? void 0 : _c.LOG_FILTER) {
        patterns = Filter.processLogFilterInput({ level: { value: level, comp: 'gte' } }, (_d = process.env) === null || _d === void 0 ? void 0 : _d.LOG_FILTER, 'environment variable LOG_FILTER.');
    }
    if (!patterns) {
        originalInput = '*';
        patterns = Filter.parseUnsafe({ level: { value: level, comp: 'gte' } }, originalInput);
    }
    return {
        // TODO looks like it actually can be undefined, bug?
        // eslint-disable-next-line
        originalInput: originalInput,
        defaults: { level: { value: level, comp: 'gte' } },
        patterns
    };
}
